File: rb-ticker/ticker.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

# Phemex WebSocket Price — subscribes to the BTCUSD 24h ticker channel,
# printing the last price to stdout on each update.
#
# Auto-reconnects on disconnect with exponential backoff (1s → 30s max).
# Sends a heartbeat (server.ping) every 20s and prints the pong to stdout.
#
# Usage:  bundle exec ruby ticker.rb

require "json"
require "websocket-client-simple"

WS_URL = "wss://ws.phemex.com"
SYMBOL = "BTCUSD"
PRICE_SCALE = 10_000
HEARTBEAT_INTERVAL = 20 # seconds
INITIAL_RECONNECT_DELAY = 1
MAX_RECONNECT_DELAY = 30

# ---------------------------------------------------------------------------
# Client
# ---------------------------------------------------------------------------

class TickerClient
  def initialize
    @ws = nil
    @last_price = nil
    @reconnect_delay = INITIAL_RECONNECT_DELAY
    @heartbeat_thread = nil
    @heartbeat_stop = false
    @running = true
    @reconnect_timer = nil
    @price_mutex = Mutex.new
  end

  def connect
    @ws&.close

    captured_ws = WebSocket::Client::Simple.connect(WS_URL)
    @ws = captured_ws
    client = self

    captured_ws.on :open do
      client.instance_variable_set(:@reconnect_delay, INITIAL_RECONNECT_DELAY)
      $stderr.puts "Connected to Phemex WebSocket"

      captured_ws.send(JSON.generate({ method: "market24h.subscribe", params: [], id: 2 }))

      client.start_heartbeat
    end

    captured_ws.on :message do |event|
      client.handle_message(event.data)
    end

    captured_ws.on :close do |_event|
      $stderr.puts "Connection closed"
      client.stop_heartbeat
      client.schedule_reconnect
    end

    captured_ws.on :error do |event|
      $stderr.puts "WebSocket error: #{event.message}" if event.message
    end
  end

  def running?
    @running
  end

  def shutdown
    @running = false
    @reconnect_timer&.kill
    stop_heartbeat
    @ws&.close
  end

  # Make these accessible from the blocks via the `client` reference
  public

  def start_heartbeat
    stop_heartbeat
    @heartbeat_stop = false
    @heartbeat_thread = Thread.new do
      until @heartbeat_stop
        sleep HEARTBEAT_INTERVAL
        break if @heartbeat_stop
        @ws&.send(JSON.generate({ method: "server.ping", params: [], id: (Time.now.to_f * 1000).to_i }))
      end
    end
  end

  def stop_heartbeat
    @heartbeat_stop = true
    @heartbeat_thread&.kill
    @heartbeat_thread = nil
  end

  def schedule_reconnect
    return unless @running

    delay = @reconnect_delay
    @reconnect_delay = [@reconnect_delay * 2, MAX_RECONNECT_DELAY].min

    $stderr.puts "Reconnecting in #{delay}s..."
    @reconnect_timer = Thread.new do
      sleep delay
      connect if @running
    end
  end

  def handle_message(raw)
    msg = JSON.parse(raw)

    # Pong — heartbeat response
    if msg["result"] == "pong"
      print "♥"
      $stdout.flush
      return
    end

    if msg["error"]
      $stderr.puts "Subscription error: #{msg["error"]}"
      return
    end

    # Ignore subscription ack
    return if msg.dig("result", "status") == "success"

    # 24h ticker channel — updates every ~1s
    market24h = msg["market24h"]
    if market24h.is_a?(Hash) && market24h["symbol"] == SYMBOL
      close_ep = market24h["close"]
      if close_ep.is_a?(Integer)
        price = close_ep.to_f / PRICE_SCALE
        log_price_if_changed(price)
      end
    end
  end

  private

  def log_price_if_changed(price)
    @price_mutex.synchronize do
      if @last_price.nil? || price != @last_price
        now = Time.now.strftime("%Y-%m-%d %H:%M:%S")
        print "\n#{now}  #{format("%.2f", price)} "
        $stdout.flush
        @last_price = price
      end
    end
  end
end

# ---------------------------------------------------------------------------
# Main
# ---------------------------------------------------------------------------

client = TickerClient.new
client.connect

Signal.trap("INT") do
  print "\nShutting down..."
  $stdout.flush
  client.shutdown
end

# Keep the main thread alive
loop do
  sleep 1
  break unless client.running?
end

puts " done"

File: py-ticker/ticker.py
#!/usr/bin/env python3
"""
Phemex WebSocket Ticker — subscribes to the BTCUSD 24h market ticker and
prints price, high, low, change %, and volume on each update.

Auto-reconnects on disconnect with exponential backoff (1s → 30s max).
Sends a heartbeat (ping) every 20s and prints a heart symbol on pong.

Usage:  python3 ticker.py
"""

import asyncio
import json
import signal
import sys
from datetime import datetime

import websockets

WS_URL = "wss://ws.phemex.com"
SYMBOL = "BTCUSD"
PRICE_SCALE = 10_000
HEARTBEAT_INTERVAL = 20  # seconds


# ---------------------------------------------------------------------------
# Ticker display
# ---------------------------------------------------------------------------

_last_price: float | None = None


def print_ticker(msg: dict) -> None:
    """Print a formatted ticker line from a market24h update message."""
    global _last_price

    fields = msg.get("market24h", {})
    if fields.get("symbol") != SYMBOL:
        return

    close_ep = fields.get("close", 0)
    open_ep = fields.get("open", 0)
    high_ep = fields.get("high", 0)
    low_ep = fields.get("low", 0)
    volume = fields.get("volume", 0)

    last = close_ep / PRICE_SCALE
    open_ = open_ep / PRICE_SCALE
    high = high_ep / PRICE_SCALE
    low = low_ep / PRICE_SCALE

    if _last_price is not None and last == _last_price:
        return  # skip unchanged price

    change_pct = ((last - open_) / open_) * 100 if open_ > 0 else 0.0
    sign = "+" if change_pct >= 0 else ""

    now = datetime.now().strftime("%m/%d/%Y, %I:%M:%S %p")
    vol_str = f"{volume:,.0f}"
    line = (
        f"{now}  {SYMBOL}  ${last:.2f}  "
        f"H: ${high:.2f}  L: ${low:.2f}  "
        f"Chg: {sign}{change_pct:.2f}%  Vol: {vol_str}"
    )
    print(line, flush=True)
    _last_price = last


# ---------------------------------------------------------------------------
# WebSocket client
# ---------------------------------------------------------------------------

_reconnect_delay = 1  # seconds, starts at 1s
MAX_RECONNECT_DELAY = 30  # seconds


async def run() -> None:
    """Connect, subscribe, and process messages, with auto-reconnect."""
    global _reconnect_delay

    while True:
        try:
            async with websockets.connect(WS_URL, ping_interval=None) as ws:
                _reconnect_delay = 1  # reset backoff on successful connect

                # Subscribe to 24h ticker channel
                subscribe_msg = json.dumps(
                    {"method": "market24h.subscribe", "params": [], "id": 2}
                )
                await ws.send(subscribe_msg)

                # Heartbeat task
                async def heartbeat():
                    while True:
                        await asyncio.sleep(HEARTBEAT_INTERVAL)
                        try:
                            await ws.send(
                                json.dumps(
                                    {
                                        "method": "server.ping",
                                        "params": [],
                                        "id": int(datetime.now().timestamp() * 1000),
                                    }
                                )
                            )
                        except websockets.ConnectionClosed:
                            break

                hb_task = asyncio.create_task(heartbeat())

                # Message loop
                async for raw in ws:
                    try:
                        msg = json.loads(raw)
                    except json.JSONDecodeError:
                        continue

                    # Pong — heartbeat response
                    if msg.get("result") == "pong":
                        print("♥", end="", flush=True)
                        continue

                    # Subscription error
                    if msg.get("error") is not None:
                        print(f"Subscription error: {msg['error']}", file=sys.stderr)
                        continue

                    # Subscription ack — ignore
                    result = msg.get("result")
                    if isinstance(result, dict) and result.get("status") == "success":
                        continue

                    # 24h ticker update
                    if "market24h" in msg:
                        print_ticker(msg)

                hb_task.cancel()
                break  # clean disconnect, don't reconnect

        except (websockets.ConnectionClosed, OSError) as exc:
            print(
                f"\nConnection lost ({exc}). Reconnecting in {_reconnect_delay}s...",
                file=sys.stderr,
            )
            await asyncio.sleep(_reconnect_delay)
            _reconnect_delay = min(_reconnect_delay * 2, MAX_RECONNECT_DELAY)


# ---------------------------------------------------------------------------
# Entry point
# ---------------------------------------------------------------------------

def main() -> None:
    try:
        asyncio.run(run())
    except KeyboardInterrupt:
        print("\nShutting down.")
        sys.exit(0)


if __name__ == "__main__":
    main()